### PR TITLE
[WIP] New multicore task executor

### PR DESCRIPTION
**Pull request by slumber**

### CODE REVIEW

File: src/multicore2.rs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+impl Worker {
+    fn spawn_in_pool(shared_state: Arc<State>) {
+        std::thread::spawn(move || {
+            let State(ref mutex, ref cv) = &*shared_state;
+            let mut guard = mutex.lock();
+
+            loop {
+                while let Some(task) = guard.0.pop_front() {
+                    drop(guard);
+                    task();
```
Comment (shamatar) [inline src/multicore2.rs:31]: What stops me to spawn infinite number of non-cooperative threads in the task?

File: src/multicore2.rs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        Worker { cpus, shared_state }
+    }
+
+    pub fn new() -> Worker {
+        Self::new_with_cpus(num_cpus::get())
+    }
+
+    pub fn compute<F, T, E>(&self, f: F) -> Receiver<Result<T, E>>
+    where
+        F: FnOnce() -> Result<T, E> + Send + 'static,
```
Comment (shamatar) [inline src/multicore2.rs:67]: The same way, F doesn't have any access to the resource constraining capabilities. It should get some "spawner", or "scope", or whatever, or at least immediately supply an number of CPU cores it wants to utilize

File: src/multicore2.rs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        let (sender, receiver) = channel::bounded(1);
+
+        let boxed_fn = Box::new(move || {
+            let result = f();
+            sender.send(result).unwrap();
+        });
+
+        {
+            let mut guard = mutex.lock();
+            guard.0.push_back(boxed_fn);
```
Comment (shamatar) [inline src/multicore2.rs:81]: You push a function that returns a result into the back of the priority queue, so a result will not become available before all the scheduled priority tasks finish, that is not an option

**Review by shamatar — CHANGES_REQUESTED** [review]

File: src/multicore2.rs (hunk 1 above)
Comment (slumber) [reply src/multicore2.rs:31]: Nothing, there're a lot of ways of halting any threadpool

File: src/multicore2.rs (hunk 2 above)
Comment (slumber) [reply src/multicore2.rs:67]: It's supposed to be used for executing subtasks, i.e. first you divide the computation into closures and then pass them to the threadpool.
We can give an access to the scope in the closure, but the problem is in FIFO task processing.
While rayon and crossbeam use local thread queues, and spawning new tasks from scope guarantees that all subtasks are going to be executed next, with our global queue this is not the case. If there're tasks awaiting in the queue, those you're spawn will screw up the order (either postponed to the queue tail or coming to the head in front of scheduled ones)

Comment (slumber) [reply src/multicore2.rs:67]: But I totally missed this use-case, my bad, will look into it

**Comment by shamatar** [comment]
What about this:
- add a function like "create tasks" that takes a closure that returns a vector of subworks
- some scoped context should be available for such closure to be able to have hierarchical structure (inherit priority, etc)
- information about potential upper limit of available resources should be passed to the "create task" too, and in general provided by worker
- each task should inform how many resources this task wants (minimal, optimal and max), and also when task is invoked it should get a number of how many resources are allocated for it
- the thread pool should also monitor a total about of resources that certain type that all the tasks in the queues want, and be clever to be able to give more than minimum if e.g. there are no other tasks that ever want that type of resource

**Comment by slumber** [comment]
> What about this:
> 
> * add a function like "create tasks" that takes a closure that returns a vector of subworks
> * some scoped context should be available for such closure to be able to have hierarchical structure (inherit priority, etc)
> * information about potential upper limit of available resources should be passed to the "create task" too, and in general provided by worker
> * each task should inform how many resources this task wants (minimal, optimal and max), and also when task is invoked it should get a number of how many resources are allocated for it
> * the thread pool should also monitor a total about of resources that certain type that all the tasks in the queues want, and be clever to be able to give more than minimum if e.g. there are no other tasks that ever want that type of resource

What kind of resources do you mean and how can we measure them? And why?
All threadpool implementations are somewhat simple since overcomplicating things often doesn't bring the best performance.
`scope` stuff is just a sugar in our case, you can directly spawn tasks without nesting it

**Pull request closed without merging**